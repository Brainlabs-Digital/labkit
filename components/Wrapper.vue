<template>
    <div :class="classes">
        <slot />
    </div>
</template>

<script>
const breakpoints = ['sm', 'md', 'lg', 'xl'];
function generateBreakpointProps(propKey) {
    const props = {};
    props[propKey] = {
        type: [Boolean, String],
    };
    breakpoints.forEach((breakpoint) => {
        props[`${breakpoint}:${propKey}`] = {
            type: [Boolean, String],
        };
    });
    return props;
}
const displayProps = generateBreakpointProps('display');
const marginTopProps = generateBreakpointProps('marginTop');
const marginBottomProps = generateBreakpointProps('marginBottom');

export default {
    computed: {
        classes() {
            const classes = [
                this.$style.base,
                ...this.generateBreakpointClasses('display', 'display'),
                ...this.generateBreakpointClasses('marginTop', 'margin-top'),
                ...this.generateBreakpointClasses(
                    'marginBottom',
                    'margin-bottom',
                ),
            ];
            return classes;
        },
    },
    methods: {
        generateBreakpointClasses(propKey, classKey) {
            const classes = [];
            if (this[propKey] === true) {
                classes.push(this.$style[classKey]);
            } else if (this[propKey] !== false) {
                classes.push(this.$style[`${classKey}--${this[propKey]}`]);
            }
            if (this[`sm:${propKey}`] === true) {
                classes.push(this.$style[`sm__${classKey}`]);
            } else if (this[`sm:${propKey}`] !== false) {
                classes.push(
                    this.$style[`sm__${classKey}--${this[`sm:${propKey}`]}`],
                );
            }
            if (this[`md:${propKey}`] === true) {
                classes.push(this.$style[`md__${classKey}`]);
            } else if (this[`md:${propKey}`] !== false) {
                classes.push(
                    this.$style[`md__${classKey}--${this[`md:${propKey}`]}`],
                );
            }
            if (this[`lg:${propKey}`] === true) {
                classes.push(this.$style[`lg__${classKey}`]);
            } else if (this[`lg:${propKey}`] !== false) {
                classes.push(
                    this.$style[`lg__${classKey}--${this[`lg:${propKey}`]}`],
                );
            }
            if (this[`xl:${propKey}`] === true) {
                classes.push(this.$style[`xl__${classKey}`]);
            } else if (this[`xl:${propKey}`] !== false) {
                classes.push(
                    this.$style[`xl__${classKey}--${this[`xl:${propKey}`]}`],
                );
            }
            return classes;
        },
    },
    props: {
        ...displayProps,
        ...marginBottomProps,
        ...marginTopProps,
    },
};
</script>

<style module lang="scss">
@import '../styles/mixins';
@import '../styles/variables';

@mixin make-classes($prefix: '') {
    .#{$prefix}display--block {
        display: block;
    }

    .#{$prefix}display--inline-block {
        display: inline-block;
    }

    .#{$prefix}display--none {
        display: none;
    }

    .#{$prefix}margin-bottom {
        margin-bottom: spacing-vertical();
    }

    .#{$prefix}margin-bottom--compact {
        margin-bottom: spacing-vertical(compact);
    }

    .#{$prefix}margin-bottom--compact-content {
        margin-bottom: spacing-vertical(compact-content);
    }

    .#{$prefix}margin-bottom--content {
        margin-bottom: spacing-vertical(content);
    }

    .#{$prefix}margin-bottom--mini {
        margin-bottom: spacing-vertical(mini);
    }

    .#{$prefix}margin-bottom--mini-content {
        margin-bottom: spacing-vertical(mini-content);
    }

    .#{$prefix}margin-bottom--none {
        margin-bottom: 0;
    }

    .#{$prefix}margin-top---base {
        margin-top: (0 - spacing-vertical());
    }

    .#{$prefix}margin-top---compact {
        margin-top: (0 - spacing-vertical(compact));
    }

    .#{$prefix}margin-top---compact-content {
        margin-top: (0 - spacing-vertical(compact-content));
    }

    .#{$prefix}margin-top---content {
        margin-top: (0 - spacing-vertical(content));
    }

    .#{$prefix}margin-top---mini {
        margin-top: (0 - spacing-vertical(mini));
    }

    .#{$prefix}margin-top---mini-content {
        margin-top: (0 - spacing-vertical(mini-content));
    }

    .#{$prefix}margin-top---none {
        margin-top: 0;
    }
}

@mixin make-breakpoint($breakpoint) {
    $breakpoint-prefix: '#{$breakpoint}__';
    @include media-breakpoint-up($breakpoint) {
        @include make-classes($breakpoint-prefix);
    }
}

.base {
    display: block;
    margin: 0;
    max-width: 100%;
    padding: 0;
}

@include make-classes();

@each $breakpoint in map-keys($grid-breakpoints) {
    @include make-breakpoint($breakpoint);
}
</style>
