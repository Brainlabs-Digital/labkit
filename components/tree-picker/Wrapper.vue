<template>
    <div>
        <lk-panel-body 
            padding="none"
            padding-bottom="none">
            <lk-navigation-tabs :items="tabs" />
        </lk-panel-body>
        <lk-panel-body
            v-if="filteredItems.length > 0 || searchTerm !== ''"
            padding-bottom="compact"
            padding-vertical="compact"
            muted
        >
            <lk-input
                v-model="searchTerm"
                margin-bottom="none"
                :placeholder="searchPlaceholder"
            >
                <template slot="prefix">
                    <lk-icon icon="search" />
                </template>
            </lk-input>
        </lk-panel-body>
        <input-error :has-error="hasError">
            <slot 
                name="error" 
                slot="error"
            />
        </input-error>
        <lk-panel-body>
            <lk-alert
                v-if="filteredItems.length === 0"
                empty
            >
                <template slot="title">{{emptyTitle}}</template>
            </lk-alert>
            <node
                v-else
                :counters="filteredCounters"
                :has-error="hasError"
                :is-expanded-tree="isExpandedTree"
                :is-root="true"
                :items="filteredItems"
                :label="rootLabel"
                :show-bulk-deselection="showBulkDeselection"
                :show-bulk-selection="showBulkSelection"
                :show-options-count="showBulkSelection"
                :show-selected-count="showBulkDeselection"
                :value="value"
                :value-set="valueSet"
                @input="emitInput"
                @toggled="toggleNode"
            />
        </lk-panel-body>
    </div>
</template>

<script>
import InputError from '../inputs/utils/InputError';
import Node from './Node';

const filters = {
    ALL: 'all',
    SELECTED: 'selected',
    UNSELECTED: 'unselected',
};
const tabs = {
    ALL: 'all',
    SELECTED: 'selected',
    UNSELECTED: 'unselected',
};

export default {
    components: {
        InputError,
        Node,
    },
    computed: {
        rootLabel() {
            if (this.filter === filters.SELECTED) {
                return `Selected ${this.optionsLabel}`;
            } else if (this.filter === filters.UNSELECTED) {
                return `Unselected ${this.optionsLabel}`;
            } else {
                return `All ${this.optionsLabel}`;
            }
        },
        counters() {
            return this.generateCounters({}, [], this.valueSet, ...this.items);
        },
        emptyTitle() {
            const optionsLabel = this.optionsLabel.toLowerCase();
            if (this.searchTerm === '') {
                if (this.filter === filters.SELECTED) {
                    return `There are no selected ${optionsLabel}`;
                } else if (this.filter === filters.UNSELECTED) {
                    return `There are no unselected ${optionsLabel}`;
                } else {
                    return `There are no ${optionsLabel}`;
                }
            } else {
                if (this.filter === filters.SELECTED) {
                    return `There are no selected ${optionsLabel} matching '${
                        this.searchTerm
                    }'`;
                } else if (this.filter === filters.UNSELECTED) {
                    return `There are no unselected ${optionsLabel} matching '${
                        this.searchTerm
                    }'`;
                } else {
                    return `There are no ${optionsLabel} matching '${
                        this.searchTerm
                    }'`;
                }
            }
        },
        filteredCounters() {
            if (this.searchTerm === '' && this.filter === filters.ALL) {
                return this.counters;
            } else {
                return this.generateCounters(
                    {},
                    [],
                    this.valueSet,
                    ...this.filteredItems,
                );
            }
        },
        filteredItems() {
            if (this.searchTerm === '' && this.filter === filters.ALL) {
                return this.items;
            } else {
                const lowerCaseSearchTerm = this.searchTerm.toLowerCase();
                const filteredItems = this.filterItems(
                    this.items,
                    this.filter,
                    lowerCaseSearchTerm,
                );
                return filteredItems;
            }
        },
        searchPlaceholder() {
            if (this.filter === filters.SELECTED) {
                return `Search Selected ${this.optionsLabel}`;
            } else if (this.filter === filters.UNSELECTED) {
                return `Search Unselected ${this.optionsLabel}`;
            } else {
                return `Search All ${this.optionsLabel}`;
            }
        },
        showBulkDeselection() {
            return this.filter !== filters.UNSELECTED;
        },
        showBulkSelection() {
            return this.filter !== filters.SELECTED;
        },
        tabs() {
            return [
                {
                    callback: () => {
                        this.setActiveTab(tabs.ALL);
                        this.setFilter(filters.ALL);
                    },
                    isActive: this.activeTab === tabs.ALL,
                    label: `All (${this.counters.options})`,
                },
                {
                    callback: () => {
                        this.setActiveTab(tabs.SELECTED);
                        this.setFilter(filters.SELECTED);
                    },
                    isActive: this.activeTab === tabs.SELECTED,
                    label: `Selected (${this.counters.selected})`,
                },
                {
                    callback: () => {
                        this.setActiveTab(tabs.UNSELECTED);
                        this.setFilter(filters.UNSELECTED);
                    },
                    isActive: this.activeTab === tabs.UNSELECTED,
                    label: `Unselected (${this.counters.options -
                        this.counters.selected})`,
                },
            ];
        },
        valueSet() {
            return new Set(this.value);
        },
    },
    data() {
        return {
            activeTab: tabs.ALL,
            filter: filters.ALL,
            isExpandedTree: {},
            rootIsExpandedTree: {},
            searchTerm: '',
            treeResetByFilter: false,
        };
    },
    methods: {
        emitInput(selectedOptions) {
            this.$emit('input', selectedOptions);
        },
        filterItems(items, filter, searchTerm) {
            /*
                We filter options only: if a parent has no options (or children that
                have options) then we filter out the parent.

                TODO:
                We may want to add a toggle option to filter by parent label, but
                we'll ignore that for now.
            */
            return items.reduce((list, item) => {
                if (
                    item.items === undefined &&
                    (searchTerm === '' ||
                        item.label.toLowerCase().includes(searchTerm) ===
                            true) &&
                    (filter !== filters.SELECTED ||
                        this.valueSet.has(item.value) === true) &&
                    (filter !== filters.UNSELECTED ||
                        this.valueSet.has(item.value) === false)
                ) {
                    list.push(item);
                } else if (item.items !== undefined) {
                    const children = this.filterItems(
                        item.items,
                        filter,
                        searchTerm,
                    );
                    if (children.length > 0) {
                        list.push({
                            ...item,
                            items: children,
                        });
                    }
                }
                return list;
            }, []);
        },
        generateCounters(counters, path, selectedItems, item, ...siblings) {
            if (item === undefined) {
                return counters;
            } else if (item.items === undefined) {
                const itemIsSelected = selectedItems.has(item.value);
                if (counters.options === undefined) {
                    counters.options = 0;
                    counters.selected = 0;
                }
                counters.options += 1;
                if (itemIsSelected === true) {
                    counters.selected += 1;
                }
                if (path.length > 0 && counters.children === undefined) {
                    counters.children = {};
                }
                let node = counters;
                for (const key of path) {
                    if (node.children === undefined) {
                        node.children = {};
                    }
                    if (node.children[key] === undefined) {
                        node.children[key] = {
                            options: 0,
                            selected: 0,
                        };
                    }
                    node = node.children[key];
                    node.options += 1;
                    if (itemIsSelected === true) {
                        node.selected += 1;
                    }
                }
            } else {
                this.generateCounters(
                    counters,
                    [...path, item.value],
                    selectedItems,
                    ...item.items,
                );
            }
            return this.generateCounters(
                counters,
                path,
                selectedItems,
                ...siblings,
            );
        },
        generateIsExpandedTree(tree, expandAll, item, ...siblings) {
            /*
                TODO:
                We want this to generate isExpanded and children at the root of
                the object.

                Currently we hack around this by supplying an initial tree with
                these values.
            */
            if (item === undefined) {
                return tree;
            } else if (item.items !== undefined) {
                tree[item.value] = {
                    isExpanded: expandAll,
                };
                const isParentOfParent = item.items.some((child) => {
                    return child.items !== undefined;
                });
                if (isParentOfParent === true) {
                    tree[item.value].children = {};
                    this.generateIsExpandedTree(
                        tree[item.value].children,
                        expandAll,
                        ...item.items,
                    );
                }
            }
            return this.generateIsExpandedTree(tree, expandAll, ...siblings);
        },
        setActiveTab(activeTab) {
            this.activeTab = activeTab;
        },
        setFilter(filter) {
            this.filter = filter;
        },
        toggleNode([path, isExpanded]) {
            let tree = this.isExpandedTree;
            for (const key of path) {
                tree = tree.children[key];
            }
            tree.isExpanded = isExpanded;
        },
    },
    props: {
        hasError: {
            default: false,
            type: Boolean,
        },
        items: {
            required: true,
            type: Array,
        },
        optionsLabel: {
            default: 'Options',
            type: String,
        },
        value: {
            default: () => [],
            type: Array,
        },
    },
    watch: {
        filter() {
            this.isExpandedTree = {
                isExpanded: false,
                children: this.generateIsExpandedTree({}, false, ...this.items),
            };
            if (this.searchTerm !== '') {
                this.searchTerm = '';
                /*
                    The reason we set rootIsExpandedTree to isExpandedTree is
                    because resetting the searchTerm triggers setting
                    isExpandedTree to rootIsExpandedTree which would then have
                    the side effect of carrying over the expanded state between
                    filters, which is not something we want.
                */
                this.rootIsExpandedTree = this.isExpandedTree;
            }
        },
        items: {
            handler() {
                this.isExpandedTree = {
                    isExpanded: false,
                    children: this.generateIsExpandedTree(
                        {},
                        false,
                        ...this.items,
                    ),
                };
            },
            immediate: true,
        },
        searchTerm(searchTerm, previousSearchTerm) {
            /*
                TODO:
                We may want to debounce this function for large amounts of
                items.

                Alternatively, if search results are greater than N we may want
                to limit the number of items that are rendered as that is where
                the bottleneck is.
                    - This should probably be a computed value (displayItems?),
                      with the limit being data that can be increased at the
                      click of a button.

                As a last resort we may need to rework some things and implement
                server side filtering and fetching.
            */
            if (searchTerm === '' && previousSearchTerm !== '') {
                this.isExpandedTree = this.rootIsExpandedTree;
                return;
            }
            if (searchTerm !== '' && previousSearchTerm === '') {
                this.rootIsExpandedTree = this.isExpandedTree;
            }
            this.isExpandedTree = {
                isExpanded: true,
                children: this.generateIsExpandedTree({}, true, ...this.items),
            };
        },
    },
};
</script>

<style module lang="scss">
@import '../../styles/variables';
</style>
