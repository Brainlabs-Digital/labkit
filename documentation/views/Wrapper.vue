<template>
    <lk-content>
        <h1>Wrapper</h1>
        <p>The wrapper component (<code>lk-wrapper</code>) is a utility componet intended to be implemented by other components that wish to provide responsive spacing and display properties.</p>
        <h3>Styling and props</h3>
        <p>By default the wrapper component will have no bottom margin and will be <code>display: block</code>.</p>
        <lk-panel margin-bottom="content">
            <lk-panel-body>
                <lk-content content-font-size-primary>
                    <p><code>display</code></p>
                    <p>The <code>display</code> prop accepts the following arguments:</p>
                    <ul>
                        <li><code>block</code>: applies styling equivalent to the CSS <code>display: block</code>.</li>
                        <li><code>inline-block</code>: applies styling equivalent to the CSS <code>display: block</code>.</li>
                        <li><code>none</code>: applies styling equivalent to the CSS <code>display: block</code>.</li>
                    </ul>
                </lk-content>
            </lk-panel-body>       
            <lk-panel-body>
                <lk-content content-font-size-primary>
                    <p><code>margin-bottom</code></p>
                    <p>The <code>margin-bottom</code> prop accepts the following arguments:</p>
                    <ul>
                        <li>With no arguments, but specified as a prop: applies styling equivalent to the CSS <code>margin-bottom: spacing-vertical()</code>.</li>
                        <li><code>compact</code> applies styling equivalent to the CSS <code>margin-bottom: spacing-vertical(compact)</code>.</li>
                        <li><code>compact-content</code> applies styling equivalent to the CSS <code>margin-bottom: spacing-vertical(compact-content)</code>.</li>
                        <li><code>content</code> applies styling equivalent to the CSS <code>margin-bottom: spacing-vertical(content)</code>.</li>
                        <li><code>mini</code> applies styling equivalent to the CSS <code>margin-bottom: spacing-vertical(mini)</code>.</li>
                        <li><code>mini-content</code> applies styling equivalent to the CSS <code>margin-bottom: spacing-vertical(mini-content)</code>.</li>
                        <li><code>none</code> applies styling equivalent to the CSS <code>margin-bottom: 0</code>.</li>
                    </ul>
                </lk-content>
            </lk-panel-body>
            <lk-panel-body>
                <lk-content content-font-size-primary>
                    <p><code>margin-top</code></p>
                    <p>The <code>margin-bottom</code> prop accepts the following arguments:</p>
                    <ul>
                        <li><code>-compact</code> applies styling equivalent to the CSS <code>margin-top: (0 - spacing-vertical(compact))</code>.</li>
                        <li><code>-compact-content</code> applies styling equivalent to the CSS <code>margin-top: (0 - spacing-vertical(compact-content))</code>.</li>
                        <li><code>-content</code> applies styling equivalent to the CSS <code>margin-top: (0 - spacing-vertical(content))</code>.</li>
                        <li><code>-mini</code> applies styling equivalent to the CSS <code>margin-top: (0 - spacing-vertical(mini))</code>.</li>
                        <li><code>-mini-content</code> applies styling equivalent to the CSS <code>margin-top: (0 - spacing-vertical(mini-content))</code>.</li>
                        <li><code>none</code> applies styling equivalent to the CSS <code>margin-top: 0</code>.</li>
                    </ul>
                </lk-content>
            </lk-panel-body>
        </lk-panel>
        <h3>Responsive breakpoints</h3>
        <p>All props of this component can be prefixed with the following prefixes to provide responsive functionality:</p>
        <ul>
            <li>Props with no prefix will apply when the browser window is >= 0px wide.</li>
            <li>Props prefixed with <code>sm:</code> will apply when the browser window is >= 576px wide.</li>
            <li>Props prefixed with <code>md:</code> will apply when the browser window is >= 768px wide.</li>
            <li>Props prefixed with <code>lg:</code> will apply when the browser window is >= 992px wide.</li>
            <li>Props prefixed with <code>xl:</code> will apply when the browser window is >= 1200px wide.</li>
        </ul>
    </lk-content>
</template>

<script>
export default {};
</script>

<style module lang="scss">
@import 'src/labkit/styles/variables';
</style>
