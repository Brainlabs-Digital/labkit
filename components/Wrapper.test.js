import '@/labkit';
import {mount} from '@vue/test-utils';
import Wrapper from './Wrapper';

/*
    NOTE:
    I'm not advocating these as "good tests": they exist to demonstrate
    snapshot testing as a concept.
*/

describe('Wrapper component', () => {
    describe('When no props are provided', () => {
        it('renders as expected', () => {
            const component = mount(Wrapper, {
                propsData: {},
            });
            expect(component.html()).toMatchSnapshot();
        });
    });

    describe('When props are provided', () => {
        it('renders as expected', () => {
            const component = mount(Wrapper, {
                propsData: {
                    display: 'none',
                    'sm:display': 'block',
                    'md:display': 'inline-block',
                },
            });
            expect(component.html()).toMatchSnapshot();
        });
    });
});
