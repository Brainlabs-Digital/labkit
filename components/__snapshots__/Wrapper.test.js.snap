// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wrapper component When no props are provided renders as expected 1`] = `<div class="base"></div>`;

exports[`Wrapper component When props are provided renders as expected 1`] = `<div class="base display--none sm__display--block md__display--inline-block"></div>`;
